aws/resource_aws_appsync_datasource.go

```diff
@@ -0,0 +1,327 @@
+		return err
+	}
+
+	d.SetId(d.Get("api_id").(string) + "-" + d.Get("name").(string))
+	d.Set("arn", resp.DataSource.DataSourceArn)
+	return nil
+}
+
+func resourceAwsAppsyncDatasourceRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appsyncconn
+
+	input := &appsync.GetDataSourceInput{
+		ApiId: aws.String(d.Get("api_id").(string)),
+		Name:  aws.String(d.Get("name").(string)),
+	}
+
+	resp, err := conn.GetDataSource(input)
+	if err != nil {
+		if isAWSErr(err, appsync.ErrCodeNotFoundException, "") {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	d.Set("arn", resp.DataSource.DataSourceArn)
+	d.Set("description", resp.DataSource.Description)
+	d.Set("dynamodb_config", flattenAppsyncDynamodbDataSourceConfig(resp.DataSource.DynamodbConfig))
+	d.Set("elasticsearch_config", flattenAppsyncElasticsearchDataSourceConfig(resp.DataSource.ElasticsearchConfig))
+	d.Set("lambda_config", flattenAppsyncLambdaDataSourceConfig(resp.DataSource.LambdaConfig))
+	d.Set("name", resp.DataSource.Name)
+	d.Set("service_role", resp.DataSource.ServiceRoleArn)
+	d.Set("type", resp.DataSource.Type)
+
+	return nil
+}
+
+func resourceAwsAppsyncDatasourceUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appsyncconn
+
+	input := &appsync.UpdateDataSourceInput{
+		ApiId: aws.String(d.Get("api_id").(string)),
+		Name:  aws.String(d.Get("name").(string)),
+	}
+
+	if d.HasChange("description") {
+		input.Description = aws.String(d.Get("description").(string))
+	}
+	if d.HasChange("service_role") {
+		input.ServiceRoleArn = aws.String(d.Get("service_role").(string))
+	}
+	if d.HasChange("type") {
+		input.Type = aws.String(d.Get("type").(string))
+	}
+
+	if d.HasChange("dynamodb_config") {
+		ddbconfig := d.Get("dynamodb_config").([]interface{})
+		if len(ddbconfig) > 0 {
+			input.DynamodbConfig = expandAppsyncDynamodbDataSourceConfig(ddbconfig[0].(map[string]interface{}))
+		}
+	}
+	if d.HasChange("elasticsearch_config") {
+		esconfig := d.Get("elasticsearch_config").([]interface{})
+		if len(esconfig) > 0 {
+			input.ElasticsearchConfig = expandAppsyncElasticsearchDataSourceConfig(esconfig[0].(map[string]interface{}))
+		}
+	}
+	if d.HasChange("lambda_config") {
+		lambdaconfig := d.Get("lambda_config").([]interface{})
+		if len(lambdaconfig) > 0 {
+			input.LambdaConfig = expandAppsyncLambdaDataSourceConfig(lambdaconfig[0].(map[string]interface{}))
+		}
+	}
+
+	_, err := conn.UpdateDataSource(input)
+	if err != nil {
+		return err
+	}
+	return resourceAwsAppsyncDatasourceRead(d, meta)
+}
+
+func resourceAwsAppsyncDatasourceDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appsyncconn
+
+	input := &appsync.DeleteDataSourceInput{
+		ApiId: aws.String(d.Get("api_id").(string)),
+		Name:  aws.String(d.Get("name").(string)),
+	}
+
+	_, err := conn.DeleteDataSource(input)
+	if err != nil {
+		if isAWSErr(err, appsync.ErrCodeNotFoundException, "") {
+			d.SetId("")
```

Comment:
The `SetId("")` is redundant here as we empty ID (and wipe the resource) in the upstream library inside Terraform's core.